File: Day023_2025/Q023_GFG.java
class Node {
    int data;
    Node next, random;

    Node(int x) {
        data = x;
        next = null;
        random = null;
    }
}

class Q023_GFG {
    public Node cloneLinkedList(Node head) {
        if (head == null) {
            return null;
        }

        // Step 1: Create new nodes and interweave them with the original list
        Node curr = head;
        while (curr != null) {
            Node newNode = new Node(curr.data);
            newNode.next = curr.next;
            curr.next = newNode;
            curr = newNode.next;
        }

        // Step 2: Assign random pointers for the newly created nodes
        curr = head;
        while (curr != null) {
            if (curr.random != null) {
                curr.next.random = curr.random.next;
            }
            curr = curr.next.next;
        }

        // Step 3: Detach the original list and the cloned list
        curr = head;
        Node newHead = curr.next;
        Node newCurr = newHead;
        while (curr != null) {
            curr.next = newCurr.next;
            curr = curr.next;
            if (curr != null) {
                newCurr.next = curr.next;
            }
            newCurr = newCurr.next;
        }

        return newHead;
    }
}

File: Day023_2025/Q023_GFG.py
class Node:
    def __init__(self, x):
        self.data = x
        self.next = None
        self.random = None

class Solution:
    def cloneLinkedList(self, head):
        if not head:
            return None

        # Step 1: Create new nodes and interweave them with the original list
        curr = head
        while curr:
            new_node = Node(curr.data)
            new_node.next = curr.next
            curr.next = new_node
            curr = new_node.next

        # Step 2: Assign random pointers for the newly created nodes
        curr = head
        while curr:
            if curr.random:
                curr.next.random = curr.random.next
            curr = curr.next.next

        # Step 3: Detach the original list and the cloned list
        curr = head
        new_head = curr.next
        new_curr = new_head
        while curr:
            curr.next = new_curr.next
            curr = curr.next
            if curr:
                new_curr.next = curr.next
            new_curr = new_curr.next

        return new_head
